packages/core/src/reducers/core.ts

```diff
@@ -120,26 +132,144 @@ const hasAjvOption = (option: any): option is InitActionOptions => {
   return false;
 };
 
+const createDynamicSchema = (
+  schema: JsonSchema,
+  uischema: UISchemaElement,
+  data: any,
+  ajv: Ajv
+): { schema: JsonSchema; updatedData: any } => {
+  if (!schema.properties) {
+    return { schema, updatedData: data };
+  }
+
+  // Start with existing required fields or empty array
+  const requiredFields = Array.isArray(schema.required)
+    ? [...schema.required]
+    : [];
+
+  // Track if anything actually changes
+  let requiredChanged = false;
+  let dataChanged = false;
+
+  // Start with references to original objects
+  let updatedData = data;
+
+  // Check each property for dynamic required rules and value rules
+  Object.entries(schema.properties).forEach(([key]) => {
+    const control = findControlForProperty(uischema, key);
+    if (control) {
+      if (hasRequiredRule(control)) {
+        // For schema-based conditions with scope: "#", we need to pass undefined as path
+        const condition = control.rule?.condition as SchemaBasedCondition;
+        const isSchemaCondition = condition && 'schema' in condition;
+        const path =
+          isSchemaCondition && condition.scope === '#' ? undefined : key;
+        const isFieldRequired = isRequired(control, data, path, ajv);
+        const fieldIndex = requiredFields.indexOf(key);
+
+        if (isFieldRequired && fieldIndex === -1) {
+          // Add to required if not already there
+          requiredFields.push(key);
+          requiredChanged = true;
+        } else if (!isFieldRequired && fieldIndex !== -1) {
+          // Remove from required if no longer required
+          requiredFields.splice(fieldIndex, 1);
+          requiredChanged = true;
+        }
+      }
+
+      if (hasValueRule(control) && data !== undefined) {
+        const condition = control.rule?.condition as SchemaBasedCondition;
+        const isSchemaCondition = condition && 'schema' in condition;
+        const path =
+          isSchemaCondition && condition.scope === '#' ? undefined : key;
+        const { shouldUpdate, newValue } = evalValue(control, data, path, ajv);
+        if (shouldUpdate) {
+          // Only create a copy if we haven't already
+          if (!dataChanged) {
+            updatedData = { ...data };
+            dataChanged = true;
+          }
+          updatedData = setFp(key, newValue, updatedData);
+        }
+      }
+
+      // Handle visibility rules - clear values for hidden fields
+      if (hasShowRule(control) && data !== undefined) {
+        const condition = control.rule?.condition as SchemaBasedCondition;
+        const isSchemaCondition = condition && 'schema' in condition;
+        const path =
+          isSchemaCondition && condition.scope === '#' ? undefined : key;
```

Comment:
Seems like this code is duplicated a few times. Probably better to extract a function instead of copy pasta